src/cryptonote_core/blockchain.cpp

```diff
@@ -3405,6 +3405,113 @@ void Blockchain::flush_invalid_blocks() {
     m_invalid_blocks.clear();
 }
 //------------------------------------------------------------------
+std::vector<eth::bls_public_key> Blockchain::get_removable_nodes() const
+{
+    // TODO: Just calculate an acceleration structure on block receive.
+    std::vector<eth::bls_public_key> bls_pubkeys_in_snl;
+    std::vector<eth::bls_public_key> bls_pubkeys_in_smart_contract;
+    {
+        // NOTE: Extract all the service nodes from the Oxen work-chain _excluding_ those that have
+        // been recently removed (e.g. voluntarily exited or dereg by protocol). Recently removed
+        // nodes on the Oxen side need to be removed from the smart contract side and this and can
+        // be done by aggregating a signature (or without if the wait time has elapsed) from the
+        // network to be removed from the smart contract (which gets re-witnessed by Oxen thereby
+        // removing it from the 'recently_removed_nodes' list).
+        auto sns = service_node_list.get_service_node_list_state();
+        bls_pubkeys_in_snl.reserve(sns.size());
+        for (const auto& sni : sns)
+            bls_pubkeys_in_snl.push_back(sni.info->bls_public_key);
+
+        // NOTE: Because of vote confirmations, we have nodes that can exist in the smart contract
+        // but not the service node list for awhile until they get confirmed at which point they get
+        // added to the SNL.
+        //
+        // We assume that these nodes will be added to the SNL. If it does, it will eventually be
+        // removed from this list and show up in the 'get_service_node_list_state()' branch above.
+        //
+        // If the node gets denied, it will be removed from this list and never added to the SNL.
+        service_node_list.for_each_pending_l2_state(
+            [&bls_pubkeys_in_snl]<typename Event>(const Event& e, const service_nodes::service_node_list::unconfirmed_l2_tx&) {
+                if constexpr  (std::is_same_v<Event, eth::event::NewServiceNode>) {
+                    bls_pubkeys_in_snl.push_back(e.bls_pubkey);
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExitRequest>) {
+                } else if constexpr (std::is_same_v<Event, eth::event::ServiceNodeExit>) {
+                } else {
+                    log::error(logcat, "Got unknown event type when iterating each pending L2 state from the SNL!");
+                }
+        });
+
+        // NOTE: Extract all service nodes from the smart contract
+        eth::RewardsContract::ServiceNodeIDs smart_contract_ids =
+                m_l2_tracker->get_all_service_node_ids(std::nullopt);
```

Comment:
Agreed strongly on updating as part of the regular L2 updating. We should contain all the (potentially unreliable/completely different network) L2 remote OPs into the L2 updating thread and avoid poisoning the codebase.

Just not sure if it should be `L2_REWARD_POOL_UPDATE_BLOCKS` or the L2 `safe_height`, I think the latter.

Reply:
The `safe_height` would be better from the perspective of just oxend, but the reason to prefer `% UPDATE_BLOCKS == 0` height comes down to keeping the number of distinct external L2 provider requests lower:
- it reduces how many calls can be made by a node (because safe_height updates every time you get the L2's height).
- it makes RPC caching (i.e. where you run a front-end cache serving multiple service nodes, with a single back-end actual L2 RPC) more effective if all of those nodes are requesting the same height.

Reply:
Oh yeah not the safe height since that's changing all the time, the `l2_height` in the block because that'd be verifiably consistent across all nodes .. though it might go out of date in certain circumstances and outside of the window of state the provider stores. 

But I like the idea that if the network stalls, everyone's view on the current state of the contract remains constant (paused at whatever state they snapped at `l2_height`, maybe they missed it entirely in which case they refuse all requests) so there's no new unknowns being introduced into the equation